mobile/android/android-application-manager.ts

```diff
@@ -44,10 +44,35 @@ export class AndroidApplicationManager extends ApplicationManagerBase {
 	}
 
 	public async startApplication(appIdentifier: string): Promise<void> {
-		await this.adb.executeShellCommand(["monkey",
-			"-p", appIdentifier,
-			"-c", "android.intent.category.LAUNCHER",
-			"1"]);
+
+		/*
+		Example "pm dump <app_identifier> | grep -A 1 MAIN" output"
+			android.intent.action.MAIN:
+			3b2df03 org.nativescript.cliapp/com.tns.NativeScriptActivity filter 50dd82e
+			Action: "android.intent.action.MAIN"
+			Category: "android.intent.category.LAUNCHER"
+			--
+			intent={act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=org.nativescript.cliapp/com.tns.NativeScriptActivity}
+			realActivity=org.nativescript.cliapp/com.tns.NativeScriptActivity
+			--
+			Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=org.nativescript.cliapp/com.tns.NativeScriptActivity }
+			frontOfTask=true task=TaskRecord{fe592ac #449 A=org.nativescript.cliapp U=0 StackId=1 sz=1}
+		*/
+		const pmDumpOutput = await this.adb.executeShellCommand(["pm", "dump", appIdentifier, "|", "grep", "-A", "1", "MAIN"]);
```

Comment:
`So what's the expected behavior when the application has more than one MAIN activity`
The first main activity will be taken. I've not seen a user case where there's more than one MAIN activity, plus the user can specify a DEFAULT main activity, but I haven't seen that either, so I didn't want to complicate the implementation if there's no need.

`Also, shouldn't we find the activity with Launcher category?`
I also thought about this, but then I found out that some part of the usual Android applications don't specify a launcher category. Read more [here](https://chris.orr.me.uk/android-apps-info-category/). Keeping this in mind I thought best not to discriminate against such apps.
Regarding the example of the two MAIN activities you provided, I haven't thought about it, because up until now, I hadn't seen this scenario, but I'm open to discussing all possible scenarios.

`Isn't it faster to call adb shell dumpsys package <appIdentifier>`
To be honest I used `pm dump` instead of `dumpsys` because when I saw the content of `pm` it seemed more reliable because it was part of the `framework.jar`. `dumpsys` command is just a binary that is somewhere in the `/system/bin` folder, and I couldn't find any information on its support and reliability.
```
generic_x86:/ # cat /system/bin/pm                                                                                                                                   
# Script to start "pm" on the device, which has a very rudimentary
# shell.
#
base=/system
export CLASSPATH=$base/framework/pm.jar
exec app_process $base/bin com.android.commands.pm.Pm "$@"

generic_x86:/ # cat /system/bin/am                                                                                                                                   
#!/system/bin/sh
#
# Script to start "am" on the device, which has a very rudimentary
# shell.
#
```
`Do we really need this grep`
I've tested this and the regex will work with the full result of the `pm dump` in case `-A 1` isn't available, but the regex works faster in a shorter amount of text so I left it with the `-A 1` option.